lib/sample_rate_converter/sample_rate_converter.c

```diff
@@ -0,0 +1,407 @@
+			return -EINVAL;
+		}
+
+		if ((output_sample_rate != 24000) && (output_sample_rate != 16000)) {
+			LOG_ERR("Invalid output sample rate for downsampling %d",
+				output_sample_rate);
+			return -EINVAL;
+		}
+	} else if (input_sample_rate < output_sample_rate) {
+		if (output_sample_rate != 48000) {
+			LOG_ERR("Invalid output sample rate for upsampling: %d",
+				output_sample_rate);
+			return -EINVAL;
+		}
+
+		if ((input_sample_rate != 24000) && (input_sample_rate != 16000)) {
+			LOG_ERR("Invalid input sample rate for upsampling: %d", input_sample_rate);
+			return -EINVAL;
+		}
+	} else {
+		LOG_ERR("Input and out samplerates are the same");
+		return -EINVAL;
+	}
+	return 0;
+}
+
+/**
+ * When upsampling from 16kHz to 48kHz the input and output bytes must be buffered. This is to
+ * fulfill the requirement for the filter that the number of input bytes must be divisible by the
+ * conversion factor. This function returns true when this is the case.
+ */
+static inline bool conversion_needs_buffering(struct sample_rate_converter_ctx *ctx,
+					      uint8_t conversion_ratio)
+{
+	return ((ctx->conversion_direction == CONVERSION_DIR_UP) && (conversion_ratio == 3));
+}
+
+static inline uint8_t calculate_conversion_ratio(uint32_t input_sample_rate,
+						 uint32_t output_sample_rate)
+{
+	if (input_sample_rate > output_sample_rate) {
+		return input_sample_rate / output_sample_rate;
+	} else {
+		return output_sample_rate / input_sample_rate;
+	}
+}
+
+/**
+ * @brief Reconfigures the sample rate converter context.
+ *
+ * @details Validates and sets all sample rate conversion parameters for the context. If buffering
+ *	    is needed for the conversion, the input buffer will be padded with two samples to
+ *	    ensure there will always be enough samples for a valid conversion.
+ *
+ * @param[in,out]	ctx			Pointer to the sample rate conversion context.
+ * @param[in]		input_sample_rate	Sample rate of the input samples.
+ * @param[in]		output_sample_rate	Sample rate of the output samples.
+ * @param[in]		filter			Filter type to use in the conversion.
+ *
+ * @retval 0 On success.
+ * @retval -EINVAL Invalid parameters used to initialize the conversion.
+ */
+static int sample_rate_converter_reconfigure(struct sample_rate_converter_ctx *ctx,
+					     uint32_t input_sample_rate,
+					     uint32_t output_sample_rate,
+					     enum sample_rate_converter_filter filter)
+{
+	int ret;
+	arm_status arm_err;
+	uint8_t *filter_coeffs;
+	size_t filter_size;
+
+	__ASSERT(ctx != NULL, "Context cannot be NULL");
+
+	ret = validate_sample_rates(input_sample_rate, output_sample_rate);
+	if (ret) {
+		LOG_ERR("Invalid sample rate given (%d)", ret);
+		return ret;
+	}
+
+	ctx->input_sample_rate = input_sample_rate;
+	ctx->output_sample_rate = output_sample_rate;
+
+	ctx->conversion_ratio = calculate_conversion_ratio(input_sample_rate, output_sample_rate);
+	ctx->conversion_direction =
```

Comment:
If input_sample_rate == output_sample_rate then CONVERSION_DIR_DOWN will be set which is a bit misleading. Suggest making a static function and handle this case.

Reply:
If `input_sample_rate == output_sample_rate`  the conversion will be rejected by `validate_sample_rates()`. So in my opinion we are already handling that case in a proper manner to ensure this case can't happen here. Do you still think we need to add another check for it?